mgr/dashboard: fix the landing page layout issues

nizamial09 opened the pull request:
![Screenshot from 2023-09-25 10-05-35](https://github.com/ceph/ceph/assets/71764184/c5d169ad-4e01-44a0-8e1e-37f91772c35c)




We were following a row-col grid layout for the landing page. First row includes Details, Status and Capacity
Second row for Inventory and Cluster Utilization

So if one of the item in the first row increases, it pushes the entire second row downwards.

To fix this, I made a col-row grid.

First col has Details and Inventory in two rows.
Second col has Status and Capacity as a col and Cluster Utilization as a single row


https://github.com/ceph/ceph/assets/71764184/079d2656-b6f1-4b63-9e54-a0fc005e03b7





**Without Alerts**
![Screenshot from 2023-09-25 10-06-05](https://github.com/ceph/ceph/assets/71764184/5e68e4a4-337a-461a-96f4-1d2a2de26df2)




**Previous one**
![Screenshot from 2023-09-19 13-13-55](https://github.com/ceph/ceph/assets/71764184/58d6c38a-6778-4d45-a53b-82e7a617d12c)





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card/card.component.ts:
```diff
@@ -14,7 +14,11 @@ export class CardComponent {
   @Input()
-  removeBorder = false;
+  borderClass = '';
   @Input()
-  shadow = false;
+  shadowClass = 'shadow-sm';
+  @Input()
+  heightClass = '';
+  @Input()
+  bgColorClass = '';
```
This is more of a philosophical discussion, but IMHO this is mixing presentational and behavioral concerns. Components (and their inputs) should care about what and how they do stuff, not what they look like.

That means that instead of having a parent component telling its children component how they have to look, it should just state what they need to do, and either:
* component look & feel is controlled externally (out-of-band) via ad-hoc CSS classes/styles.
* components adapt themselves their look & feel according to functional/controlled inputs (I'd prefer the `removeBorder`, or even better, just `border`, even if they cope with presentational concerns, than letting it pass a CSS class).

Classes shouldn't allow input arguments that can potentially break them, or alter their look & feel in undefined ways (and passing class names allows both things).

cloudbehl commented:
@nizamial09 Thanks, this overall looks better.

Few things to take care:
1. Alert panel has scrolls bot vertical and horizontal, we could remove horizontal totally and vertical with 2 alerts.
2. The icon for warning and danger can be center aligned and smaller in size.
3. Pie chart looks so big, we can add padding on left to make it look smaller and better. 

![Screenshot from 2023-09-25 14-27-42](https://github.com/ceph/ceph/assets/6695156/edf9af5f-d2ba-4b45-a078-904294779cfb)

Pegonzal reviewed (commented):
Looks pretty good to me, my feedback is on the same topics that have already been mentioned:

On the status card:

- I think we can get rid of the horizontal bar, and make use only of the vertical one.
- The last alert should go all the way to the bottom of the card, to avoid this wierd space here
![image](https://github.com/ceph/ceph/assets/106810838/bbb2cf2b-5315-420e-9ce1-cf6a93bf830c)

- If both alerts are being shown by defualt, should the alerts badges be "activated"?
![image](https://github.com/ceph/ceph/assets/106810838/e2f5bde7-31bb-40f3-b5f9-f7260db784f1)


Layout:

Looks like on larger screens the layout goes a bit wild, the capacity card should follow the statu's card height and the cluster utilization card should fit accordingly to their height.
Currently the capacity card will go over the status card height
![image](https://github.com/ceph/ceph/assets/106810838/15b2c7f8-14e4-46bd-8014-1673a06b8f35)

cloudbehl commented:
On capacity card, can we fix the size of donut and legend? 

[Screencast from 2023-09-26 11-17-29.webm](https://github.com/ceph/ceph/assets/6695156/90b7654b-f046-4f65-b55c-3f436a23c7d4)

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:
```diff
@@ -64,18 +64,18 @@ export class PrometheusAlertService {
       this.notifyOnAlertChanges(alerts, this.alerts);
     }
     this.activeAlerts = _.reduce<AlertmanagerAlert, number>(
-      this.alerts,
+      alerts,
       (result, alert) => (alert.status.state === 'active' ? ++result : result),
       0
     );
```
Looks like active alerts are not sorted right now, can you do that as well?

nizamial09 replied:
how would you want me to sort it? Recent ones first or the `critical first and warnings later`. As a user, I think I'd prefer to have the recent ones shown to me first. Or I prefer to have a control for the filtering. Either recent or the type of alert.

cloudbehl replied:
Can we sort it via the recent critical alerts?

nizamial09 replied:
but that would mean, the recent warnings are lost below in the list right? Atleast in the Status card where there is no way to sort and force the user to scroll down?

nizamial09 replied:
maybe i can sort it with recent alerts for now in the status card and later on create a toggle to sort critical-warning alerts?

cloudbehl replied:
For recent warning , user has options to filter by the warning and have look if they want.  By clicking on warning they can have a look at recent warnings anytime. As warning holds less importance when they have critical alerts.

nizamial09 replied:
if you are okay, i'd like to take this up in a different PR, preferrably something like this: https://github.com/ceph/ceph/pull/53679.

I was wondering about the sorting and the processing we have to do in the front-end. Then I saw the grouping in the prometheus alerts. So thought this could be nicer if we can do it from the backend rather than processing it in the frontend.

cloudbehl replied:
Ideally best to be done from backend.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-pie/dashboard-pie.component.ts:
```diff
@@ -45,6 +45,8 @@ export class DashboardPieComponent implements OnChanges, OnInit {
     options: {
       cutoutPercentage: 70,
+      responsive: true,
+      maintainAspectRatio: false,
```
done!

nizamial09 commented:
updated page looks like this. 

@cloudbehl, @Pegonzal : adding margins to the card will not look good. Also when the alerts are not there I need to maintain the original size which is shrinked and misaligned if the padding is added manually each time. So I prefer center aligning here and keeping the original size (even though that also creates an empty spaces around the card). 
![Screenshot from 2023-09-26 12-35-28](https://github.com/ceph/ceph/assets/71764184/7861d0de-f1b8-486d-9e59-eadcd71ae499)

nizamial09 commented:
> * I think we can get rid of the horizontal bar, and make use only of the vertical one.
> *

I didn't understood it? which horizontal bar where you mentioning. Is it the scroll bar. if so, its removed now.

> * If both alerts are being shown by defualt, should the alerts badges be "activated"?

That would cause confusion I guess. Let's say if both the pills are active, then which one you click to show just the `critical` ones? I guess this is better. If none is clicked, it shows all. Otherwise shows whatever is filtered. This way it acts like a filter toggle!

nizamial09 commented:
> > * I think we can get rid of the horizontal bar, and make use only of the vertical one.
> 
> I didn't understood it? which horizontal bar where you mentioning. Is it the scroll bar. if so, its removed now.
> 
> > * If both alerts are being shown by defualt, should the alerts badges be "activated"?
> 
> That would cause confusion I guess. Let's say if both the pills are active, then which one you click to show just the `critical` ones? I guess this is better. If none is clicked, it shows all. Otherwise shows whatever is filtered. This way it acts like a filter toggle!

> Looks like on larger screens the layout goes a bit wild, the capacity card should follow the statu's card height and the cluster utilization card should fit accordingly to their height.
> Currently the capacity card will go over the status card height

I haven't seen this though in my UWD monitor. can you check again?

nizamial09 commented:
@cloudbehl addressed your comments as well

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -1,296 +1,314 @@
+              <div class="alerts-section pt-0">
+                <hr class="mt-1 mb-1">
+                <ngx-simplebar [options]="simplebar">
+                  <div class="card-body ps-0 pe-0 pt-1 pb-0">
```
```suggestion
                  <div class="card-body ps-0 pe-0 pt-0 pb-0">
```

nizamial09 replied:
done, thanks

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -1,296 +1,314 @@
+              </div>
+
+              <div class="alerts-section pt-0">
+                <hr class="mt-1 mb-1">
```
```suggestion
                <hr class="mt-1 mb-0">
```

nizamial09 replied:
done!

cloudbehl reviewed (commented):
We can remove `pb-2 pt-2` from the inventory, as it adds so much space in each row,

https://github.com/ceph/ceph/blob/98fe7940c8496dfdbd462236f0354f23a9896ac9/src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html#L1

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -1,296 +1,314 @@
+      <!-- Inventory Card-->
+      <cd-card cardTitle="Inventory"
+               i18n-title
+               class="pt-4 mb-4"
```
Adds extra padding could be removed.
```suggestion
               class="pt-4"
```

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -1,296 +1,314 @@
+            <div class="card-body ps-0 pe-1 pb-1 pt-0">
               <h6 class="card-title bold">{{ alert.labels.alertname }}</h6>
               <p class="card-text me-3"
                  [innerHtml]="alert.annotations.summary"></p>
```
Lets show description of alert instead of summary.
```suggestion
                 [innerHtml]="alert.annotations.description"></p>
```

nizamial09 replied:
hmm, summary would be better for an overview page as description is more descriptive and doesn't look nicer with the view. 
![Screenshot from 2023-09-26 21-28-33](https://github.com/ceph/ceph/assets/71764184/ec16a7f5-774f-48f9-ad3e-c8656c526573)

cloudbehl replied:
The problem I see with Summary is it doesn't adds value, the messages in summary are pretty vague.

IMO nicer view vs value addition, I will choose both. What we can do is to add text ellipsis if the text is too huge by setting the max height if each alert row. If you think it doesn't look good.

I think it just looks fine with large test as well.

cloudbehl commented:
Status need to center align back when we don't have alerts.


![image](https://github.com/ceph/ceph/assets/6695156/b06a92c2-0207-4f83-95f5-842f3c36d018)

nizamial09 commented:
> We can remove `pb-2 pt-2` from the inventory, as it adds so much space in each row,

Personally, I prefer to have that space inorder to make the card fill the view. Also the empty spaces are not much noticable with the contents and checkboxes inside. Maybe in your inventory refactor PR, we can try and remove that spaces. And currently, as it is, the padding matches the one in the [patternfly design](https://www.patternfly.org/patterns/dashboard/design-guidelines/#inventory-card), (not exactly accurately but to some level)

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/prometheus-alert.service.ts:
```diff
@@ -64,18 +64,18 @@ export class PrometheusAlertService {
+      alerts,
       (result, alert) =>
         alert.status.state === 'active' && alert.labels.severity === 'warning' ? ++result : result,
       0
```
Probably a separate PR would be better for this!

cloudbehl replied:
Can you create tracker

nizamial09 replied:
https://tracker.ceph.com/issues/62990

nizamial09 commented:
> Status need to center align back when we don't have alerts.
> 
> ![image](https://user-images.githubusercontent.com/6695156/270658120-b06a92c2-0207-4f83-95f5-842f3c36d018.png)

@cloudbehl i've been trying to make it center for sometime now and miserably failed. (i've started remember the trolls where a developer tries endlessly to center a div). could you help me out here?

cloudbehl reviewed (approved):
LGTM

Pegonzal reviewed (approved):
Looks really good, thanks @nizamial09!

For larger screens it looks like it is still breaking a bit, would be nice to have the capacity card follow the status's card height, but I would consider this a nit and not needed for this PR to go in

![image](https://github.com/ceph/ceph/assets/106810838/1a833e04-44a4-4b70-af16-cd844db8afb1)

nizamial09 commented:
> For larger screens it looks like it is still breaking a bit

Thanks @Pegonzal i thought i fixed it locally but checking the code again i forgot to push the actual changes which fixed this particular issue. Updated the PR, this should be fixed now!
![image](https://github.com/ceph/ceph/assets/71764184/feb196d9-2143-4f53-baf1-8b5128757234)
